Support sycl::select for bool

**Describe the bug**
[Table 178](https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#table.relational.functions.marray) from the SYCL 2020 specification defines `sycl::select` as follows:

> ```c++
> gentype select (gentype a, gentype b, genbool c)
> ```
> Returns the component-wise `result = c ? b : a`.

where `gentype` is any scalar or `marray` type and `genbool` is a scalar `bool` or an `marray<bool>`.  https://github.com/KhronosGroup/SYCL-Docs/pull/234 clarifies that `select` is intended to support scalar types.

The DPC++ compiler does not support scalar types, and exits with the following error:
```
select.cpp:10:20: error: no matching function for call to 'select'
    float result = sycl::select(a, b, c);
                   ^~~~~~~~~~~~
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1313:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<char>, sycl::detail::type_list<signed char>>, sycl::detail::type_list<short>, sycl::detail::type_list<int>, sycl::detail::type_list<long>, sycl::detail::type_list<long long>>, sycl::detail::type_list<sycl::detail::type_list<unsigned char>, sycl::detail::type_list<unsigned short>, sycl::detail::type_list<unsigned int>, sycl::detail::type_list<unsigned long>, sycl::detail::type_list<unsigned long long>, sycl::detail::type_list<std::byte>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::vec<char, 1>, sycl::vec<char, 2>, sycl::vec<char, 3>, sycl::vec<char, 4>, sycl::vec<char, 8>, sycl::vec<char, 16>>, sycl::detail::type_list<sycl::vec<signed char, 1>, sycl::vec<signed char, 2>, sycl::vec<signed char, 3>, sycl::vec<signed char, 4>, sycl::vec<signed char, 8>, sycl::vec<signed char, 16>>>, sycl::detail::type_list<sycl::vec<short, 1>, sycl::vec<short, 2>, sycl::vec<short, 3>, sycl::vec<short, 4>, sycl::vec<short, 8>, sycl::vec<short, 16>>, sycl::detail::type_list<sycl::vec<int, 1>, sycl::vec<int, 2>, sycl::vec<int, 3>, sycl::vec<int, 4>, sycl::vec<int, 8>, sycl::vec<int, 16>>, sycl::detail::type_list<sycl::vec<long, 1>, sycl::vec<long, 2>, sycl::vec<long, 3>, sycl::vec<long, 4>, sycl::vec<long, 8>, sycl::vec<long, 16>>, sycl::detail::type_list<sycl::vec<long long, 1>, sycl::vec<long long, 2>, sycl::vec<long long, 3>, sycl::vec<long long, 4>, sycl::vec<long long, 8>, sycl::vec<long long, 16>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::vec<unsigned char, 1>, sycl::vec<unsigned char, 2>, sycl::vec<unsigned char, 3>, sycl::vec<unsigned char, 4>, sycl::vec<unsigned char, 8>, sycl::vec<unsigned char, 16>>, sycl::detail::type_list<sycl::vec<unsigned short, 1>, sycl::vec<unsigned short, 2>, sycl::vec<unsigned short, 3>, sycl::vec<unsigned short, 4>, sycl::vec<unsigned short, 8>, sycl::vec<unsigned short, 16>>, sycl::detail::type_list<sycl::vec<unsigned int, 1>, sycl::vec<unsigned int, 2>, sycl::vec<unsigned int, 3>, sycl::vec<unsigned int, 4>, sycl::vec<unsigned int, 8>, sycl::vec<unsigned int, 16>>, sycl::detail::type_list<sycl::vec<unsigned long, 1>, sycl::vec<unsigned long, 2>, sycl::vec<unsigned long, 3>, sycl::vec<unsigned long, 4>, sycl::vec<unsigned long, 8>, sycl::vec<unsigned long, 16>>, sycl::detail::type_list<sycl::vec<unsigned long long, 1>, sycl::vec<unsigned long long, 2>, sycl::vec<unsigned long long, 3>, sycl::vec<unsigned long long, 4>, sycl::vec<unsigned long long, 8>, sycl::vec<unsigned long long, 16>>, sycl::detail::type_list<sycl::vec<std::byte, 1>, sycl::vec<std::byte, 2>, sycl::vec<std::byte, 3>, sycl::vec<std::byte, 4>, sycl::vec<std::byte, 8>, sycl::vec<std::byte, 16>>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::marray<char, 1>, sycl::marray<char, 2>, sycl::marray<char, 3>, sycl::marray<char, 4>, sycl::marray<char, 8>, sycl::marray<char, 16>>, sycl::detail::type_list<sycl::marray<signed char, 1>, sycl::marray<signed char, 2>, sycl::marray<signed char, 3>, sycl::marray<signed char, 4>, sycl::marray<signed char, 8>, sycl::marray<signed char, 16>>>, sycl::detail::type_list<sycl::marray<short, 1>, sycl::marray<short, 2>, sycl::marray<short, 3>, sycl::marray<short, 4>, sycl::marray<short, 8>, sycl::marray<short, 16>>, sycl::detail::type_list<sycl::marray<int, 1>, sycl::marray<int, 2>, sycl::marray<int, 3>, sycl::marray<int, 4>, sycl::marray<int, 8>, sycl::marray<int, 16>>, sycl::detail::type_list<sycl::marray<long, 1>, sycl::marray<long, 2>, sycl::marray<long, 3>, sycl::marray<long, 4>, sycl::marray<long, 8>, sycl::marray<long, 16>>, sycl::detail::type_list<sycl::marray<long long, 1>, sycl::marray<long long, 2>, sycl::marray<long long, 3>, sycl::marray<long long, 4>, sycl::marray<long long, 8>, sycl::marray<long long, 16>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::marray<unsigned char, 1>, sycl::marray<unsigned char, 2>, sycl::marray<unsigned char, 3>, sycl::marray<unsigned char, 4>, sycl::marray<unsigned char, 8>, sycl::marray<unsigned char, 16>>, sycl::detail::type_list<sycl::marray<unsigned short, 1>, sycl::marray<unsigned short, 2>, sycl::marray<unsigned short, 3>, sycl::marray<unsigned short, 4>, sycl::marray<unsigned short, 8>, sycl::marray<unsigned short, 16>>, sycl::detail::type_list<sycl::marray<unsigned int, 1>, sycl::marray<unsigned int, 2>, sycl::marray<unsigned int, 3>, sycl::marray<unsigned int, 4>, sycl::marray<unsigned int, 8>, sycl::marray<unsigned int, 16>>, sycl::detail::type_list<sycl::marray<unsigned long, 1>, sycl::marray<unsigned long, 2>, sycl::marray<unsigned long, 3>, sycl::marray<unsigned long, 4>, sycl::marray<unsigned long, 8>, sycl::marray<unsigned long, 16>>, sycl::detail::type_list<sycl::marray<unsigned long long, 1>, sycl::marray<unsigned long long, 2>, sycl::marray<unsigned long long, 3>, sycl::marray<unsigned long long, 4>, sycl::marray<unsigned long long, 8>, sycl::marray<unsigned long long, 16>>, sycl::detail::type_list<sycl::marray<std::byte, 1>, sycl::marray<std::byte, 2>, sycl::marray<std::byte, 3>, sycl::marray<std::byte, 4>, sycl::marray<std::byte, 8>, sycl::marray<std::byte, 16>>>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1322:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<char>, sycl::detail::type_list<signed char>>, sycl::detail::type_list<short>, sycl::detail::type_list<int>, sycl::detail::type_list<long>, sycl::detail::type_list<long long>>, sycl::detail::type_list<sycl::detail::type_list<unsigned char>, sycl::detail::type_list<unsigned short>, sycl::detail::type_list<unsigned int>, sycl::detail::type_list<unsigned long>, sycl::detail::type_list<unsigned long long>, sycl::detail::type_list<std::byte>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::vec<char, 1>, sycl::vec<char, 2>, sycl::vec<char, 3>, sycl::vec<char, 4>, sycl::vec<char, 8>, sycl::vec<char, 16>>, sycl::detail::type_list<sycl::vec<signed char, 1>, sycl::vec<signed char, 2>, sycl::vec<signed char, 3>, sycl::vec<signed char, 4>, sycl::vec<signed char, 8>, sycl::vec<signed char, 16>>>, sycl::detail::type_list<sycl::vec<short, 1>, sycl::vec<short, 2>, sycl::vec<short, 3>, sycl::vec<short, 4>, sycl::vec<short, 8>, sycl::vec<short, 16>>, sycl::detail::type_list<sycl::vec<int, 1>, sycl::vec<int, 2>, sycl::vec<int, 3>, sycl::vec<int, 4>, sycl::vec<int, 8>, sycl::vec<int, 16>>, sycl::detail::type_list<sycl::vec<long, 1>, sycl::vec<long, 2>, sycl::vec<long, 3>, sycl::vec<long, 4>, sycl::vec<long, 8>, sycl::vec<long, 16>>, sycl::detail::type_list<sycl::vec<long long, 1>, sycl::vec<long long, 2>, sycl::vec<long long, 3>, sycl::vec<long long, 4>, sycl::vec<long long, 8>, sycl::vec<long long, 16>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::vec<unsigned char, 1>, sycl::vec<unsigned char, 2>, sycl::vec<unsigned char, 3>, sycl::vec<unsigned char, 4>, sycl::vec<unsigned char, 8>, sycl::vec<unsigned char, 16>>, sycl::detail::type_list<sycl::vec<unsigned short, 1>, sycl::vec<unsigned short, 2>, sycl::vec<unsigned short, 3>, sycl::vec<unsigned short, 4>, sycl::vec<unsigned short, 8>, sycl::vec<unsigned short, 16>>, sycl::detail::type_list<sycl::vec<unsigned int, 1>, sycl::vec<unsigned int, 2>, sycl::vec<unsigned int, 3>, sycl::vec<unsigned int, 4>, sycl::vec<unsigned int, 8>, sycl::vec<unsigned int, 16>>, sycl::detail::type_list<sycl::vec<unsigned long, 1>, sycl::vec<unsigned long, 2>, sycl::vec<unsigned long, 3>, sycl::vec<unsigned long, 4>, sycl::vec<unsigned long, 8>, sycl::vec<unsigned long, 16>>, sycl::detail::type_list<sycl::vec<unsigned long long, 1>, sycl::vec<unsigned long long, 2>, sycl::vec<unsigned long long, 3>, sycl::vec<unsigned long long, 4>, sycl::vec<unsigned long long, 8>, sycl::vec<unsigned long long, 16>>, sycl::detail::type_list<sycl::vec<std::byte, 1>, sycl::vec<std::byte, 2>, sycl::vec<std::byte, 3>, sycl::vec<std::byte, 4>, sycl::vec<std::byte, 8>, sycl::vec<std::byte, 16>>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::detail::type_list<sycl::marray<char, 1>, sycl::marray<char, 2>, sycl::marray<char, 3>, sycl::marray<char, 4>, sycl::marray<char, 8>, sycl::marray<char, 16>>, sycl::detail::type_list<sycl::marray<signed char, 1>, sycl::marray<signed char, 2>, sycl::marray<signed char, 3>, sycl::marray<signed char, 4>, sycl::marray<signed char, 8>, sycl::marray<signed char, 16>>>, sycl::detail::type_list<sycl::marray<short, 1>, sycl::marray<short, 2>, sycl::marray<short, 3>, sycl::marray<short, 4>, sycl::marray<short, 8>, sycl::marray<short, 16>>, sycl::detail::type_list<sycl::marray<int, 1>, sycl::marray<int, 2>, sycl::marray<int, 3>, sycl::marray<int, 4>, sycl::marray<int, 8>, sycl::marray<int, 16>>, sycl::detail::type_list<sycl::marray<long, 1>, sycl::marray<long, 2>, sycl::marray<long, 3>, sycl::marray<long, 4>, sycl::marray<long, 8>, sycl::marray<long, 16>>, sycl::detail::type_list<sycl::marray<long long, 1>, sycl::marray<long long, 2>, sycl::marray<long long, 3>, sycl::marray<long long, 4>, sycl::marray<long long, 8>, sycl::marray<long long, 16>>>, sycl::detail::type_list<sycl::detail::type_list<sycl::marray<unsigned char, 1>, sycl::marray<unsigned char, 2>, sycl::marray<unsigned char, 3>, sycl::marray<unsigned char, 4>, sycl::marray<unsigned char, 8>, sycl::marray<unsigned char, 16>>, sycl::detail::type_list<sycl::marray<unsigned short, 1>, sycl::marray<unsigned short, 2>, sycl::marray<unsigned short, 3>, sycl::marray<unsigned short, 4>, sycl::marray<unsigned short, 8>, sycl::marray<unsigned short, 16>>, sycl::detail::type_list<sycl::marray<unsigned int, 1>, sycl::marray<unsigned int, 2>, sycl::marray<unsigned int, 3>, sycl::marray<unsigned int, 4>, sycl::marray<unsigned int, 8>, sycl::marray<unsigned int, 16>>, sycl::detail::type_list<sycl::marray<unsigned long, 1>, sycl::marray<unsigned long, 2>, sycl::marray<unsigned long, 3>, sycl::marray<unsigned long, 4>, sycl::marray<unsigned long, 8>, sycl::marray<unsigned long, 16>>, sycl::detail::type_list<sycl::marray<unsigned long long, 1>, sycl::marray<unsigned long long, 2>, sycl::marray<unsigned long long, 3>, sycl::marray<unsigned long long, 4>, sycl::marray<unsigned long long, 8>, sycl::marray<unsigned long long, 16>>, sycl::detail::type_list<sycl::marray<std::byte, 1>, sycl::marray<std::byte, 2>, sycl::marray<std::byte, 3>, sycl::marray<std::byte, 4>, sycl::marray<std::byte, 8>, sycl::marray<std::byte, 16>>>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1331:1: note: candidate template ignored: requirement 'detail::is_contained<bool, sycl::detail::type_list<sycl::detail::type_list<int>, sycl::detail::type_list<sycl::vec<int, 1>, sycl::vec<int, 2>, sycl::vec<int, 3>, sycl::vec<int, 4>, sycl::vec<int, 8>, sycl::vec<int, 16>>, sycl::detail::type_list<sycl::marray<int, 1>, sycl::marray<int, 2>, sycl::marray<int, 3>, sycl::marray<int, 4>, sycl::marray<int, 8>, sycl::marray<int, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1340:1: note: candidate template ignored: requirement 'detail::is_contained<bool, sycl::detail::type_list<sycl::detail::type_list<unsigned int>, sycl::detail::type_list<sycl::vec<unsigned int, 1>, sycl::vec<unsigned int, 2>, sycl::vec<unsigned int, 3>, sycl::vec<unsigned int, 4>, sycl::vec<unsigned int, 8>, sycl::vec<unsigned int, 16>>, sycl::detail::type_list<sycl::marray<unsigned int, 1>, sycl::marray<unsigned int, 2>, sycl::marray<unsigned int, 3>, sycl::marray<unsigned int, 4>, sycl::marray<unsigned int, 8>, sycl::marray<unsigned int, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1349:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<double>, sycl::detail::type_list<sycl::vec<double, 1>, sycl::vec<double, 2>, sycl::vec<double, 3>, sycl::vec<double, 4>, sycl::vec<double, 8>, sycl::vec<double, 16>>, sycl::detail::type_list<sycl::marray<double, 1>, sycl::marray<double, 2>, sycl::marray<double, 3>, sycl::marray<double, 4>, sycl::marray<double, 8>, sycl::marray<double, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1358:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<double>, sycl::detail::type_list<sycl::vec<double, 1>, sycl::vec<double, 2>, sycl::vec<double, 3>, sycl::vec<double, 4>, sycl::vec<double, 8>, sycl::vec<double, 16>>, sycl::detail::type_list<sycl::marray<double, 1>, sycl::marray<double, 2>, sycl::marray<double, 3>, sycl::marray<double, 4>, sycl::marray<double, 8>, sycl::marray<double, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1367:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::half_impl::half>, sycl::detail::type_list<sycl::vec<sycl::detail::half_impl::half, 1>, sycl::vec<sycl::detail::half_impl::half, 2>, sycl::vec<sycl::detail::half_impl::half, 3>, sycl::vec<sycl::detail::half_impl::half, 4>, sycl::vec<sycl::detail::half_impl::half, 8>, sycl::vec<sycl::detail::half_impl::half, 16>>, sycl::detail::type_list<sycl::marray<sycl::detail::half_impl::half, 1>, sycl::marray<sycl::detail::half_impl::half, 2>, sycl::marray<sycl::detail::half_impl::half, 3>, sycl::marray<sycl::detail::half_impl::half, 4>, sycl::marray<sycl::detail::half_impl::half, 8>, sycl::marray<sycl::detail::half_impl::half, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
/llvm/build//bin/../include/sycl/CL/sycl/builtins.hpp:1376:1: note: candidate template ignored: requirement 'detail::is_contained<float, sycl::detail::type_list<sycl::detail::type_list<sycl::detail::half_impl::half>, sycl::detail::type_list<sycl::vec<sycl::detail::half_impl::half, 1>, sycl::vec<sycl::detail::half_impl::half, 2>, sycl::vec<sycl::detail::half_impl::half, 3>, sycl::vec<sycl::detail::half_impl::half, 4>, sycl::vec<sycl::detail::half_impl::half, 8>, sycl::vec<sycl::detail::half_impl::half, 16>>, sycl::detail::type_list<sycl::marray<sycl::detail::half_impl::half, 1>, sycl::marray<sycl::detail::half_impl::half, 2>, sycl::marray<sycl::detail::half_impl::half, 3>, sycl::marray<sycl::detail::half_impl::half, 4>, sycl::marray<sycl::detail::half_impl::half, 8>, sycl::marray<sycl::detail::half_impl::half, 16>>>>::value' was not satisfied [with T = float, T2 = bool]
select(T a, T b, T2 c) __NOEXC {
^
1 error generated.
```

**To Reproduce**
```c++
#include <sycl/sycl.hpp>

int main()
{
  sycl::queue q;
  q.single_task([=]() {
    bool c = true;
    float a = 1.0f;
    float b = 2.0f;
    float result = sycl::select(a, b, c);
  });
}
```

Compile with:
`clang++ -fsycl select.cpp`

**Environment:**

- OS: All
- Target device and vendor: All
- DPC++ version: 233116040d8236bd2d1321474444878a7159a98d
- Dependencies version: N/A

**Additional context**
I'd expect this to lower to `OpSelect` in SPIR-V and a ternary otherwise.

Although this is a very simple feature, it's very important because it enables developers to write this sort of branch-free code in a (performance) portable way.  Without this fix, developers have to use a ternary, an `if`/`else` block or a complicated arithmetic expression -- any of these workarounds are sufficient for correctness, but different compilers/backends/devices will have different preferences and performance characteristics.

--- a/sycl/include/CL/sycl/builtins.hpp
+++ b/sycl/include/CL/sycl/builtins.hpp
@@ -1306,6 +1306,13 @@ detail::enable_if_t<detail::is_gentype<T>::value, T> bitselect(T a, T b,
   return __sycl_std::__invoke_bitselect<T>(a, b, c);
 }
 
+// sgentype select (sgentype a, sgentype b, bool c)
+template <typename T>
+detail::enable_if_t<detail::is_sgentype<T>::value, T> select(T a, T b,
+                                                             bool c) __NOEXC {
+  return __sycl_std::__invoke_select<T>(a, b, (int) c);
+}
+
 // geninteger select (geninteger a, geninteger b, igeninteger c)
 template <typename T, typename T2>
 detail::enable_if_t<

--- a/sycl/include/CL/sycl/builtins.hpp
+++ b/sycl/include/CL/sycl/builtins.hpp
@@ -1310,7 +1310,7 @@ detail::enable_if_t<detail::is_gentype<T>::value, T> bitselect(T a, T b,
 template <typename T>
 detail::enable_if_t<detail::is_sgentype<T>::value, T> select(T a, T b,
                                                              bool c) __NOEXC {
-  return __sycl_std::__invoke_select<T>(a, b, (int) c);
+  return __sycl_std::__invoke_select<T>(a, b, static_cast<int>(c));
 }
 
 // geninteger select (geninteger a, geninteger b, igeninteger c)

